File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<InputType, OutputType>(
  items: InputType[],
  predicate: (item: InputType) => boolean,
  consumer: (item: InputType) => OutputType
): OutputType | undefined {
  const foundItem = items.find(predicate);
  return foundItem !== undefined ? consumer(foundItem) : undefined;
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  for (let power = 1n; ; power *= base) {
    yield power
  }
}

export async function meaningfulLineCount(filename: string): Promise<number> {
  let count = 0;
  const file = await open(filename, "r")
  for await (const line of file.readLines()) {
    const trimmed = line.trim()
    if (trimmed && !trimmed.startsWith("#")) {
      count++
    }
  }
  return count
}

interface Sphere {
  kind: "Sphere"
  radius: number
}

interface Box {
  kind: "Box"
  width: number
  height: number
  depth: number
}

export type Shape = Sphere | Box

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return 2 * (shape.width * shape.height + shape.width * shape.depth + shape.height * shape.depth);
    case "Sphere":
      return 4 * Math.PI * Math.pow(shape.radius, 2);
  }
}

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return shape.width * shape.height * shape.depth;
    case "Sphere":
      return (4 / 3) * Math.PI * Math.pow(shape.radius, 3);
  }
}

export function shapeToString(shape: Shape): string {
  switch (shape.kind) {
    case "Box":
      return `Box(width=${shape.width}, length=${shape.height}, depth=${shape.depth})`;
    case "Sphere":
      return `Sphere(radius=${shape.radius})`;
  }
}

// singleton?

interface Comparable<T> {
  compareTo(other: T): number;
}
interface BinarySearchTree<T extends Comparable<T>> {
  insert(value: T): BinarySearchTree<T>;
  contains(value: T): boolean;
  size(): number;
  inorder(): IterableIterator<T>;
}
class Empty<T extends Comparable<T>> implements BinarySearchTree<T> {
  static instance = new Empty<any>();

  insert(value: T): BinarySearchTree<T> {
    return new Node(value, Empty.instance, Empty.instance);
  }

  contains(_: T): boolean { return false; }

  size(): number { return 0; }

  *inorder(): IterableIterator<T> {}

  toString(): string { return "()"; }
}

class Node<T extends Comparable<T>> implements BinarySearchTree<T> {
  constructor(
    private readonly value: T,
    private readonly left: BinarySearchTree<T>,
    private readonly right: BinarySearchTree<T>
  ) {}

  insert(value: T): BinarySearchTree<T> {
    if (value.compareTo(this.value) < 0) {
      return new Node(this.value, this.left.insert(value), this.right);
    } else if (value.compareTo(this.value) > 0) {
      return new Node(this.value, this.left, this.right.insert(value));
    }
    return this; 
  }

  contains(value: T): boolean {
    if (value.compareTo(this.value) < 0) {
      return this.left.contains(value);
    } else if (value.compareTo(this.value) > 0) {
      return this.right.contains(value);
    }
    return true;
  }

  size(): number {
    return 1 + this.left.size() + this.right.size();
  }

  *inorder(): IterableIterator<T> {
    yield* this.left.inorder();
    yield this.value;
    yield* this.right.inorder();
  }

  toString(): string {
    return `(${this.left}${this.value}${this.right})`;
  }
}

class ComparableString implements Comparable<ComparableString> {
  constructor(private readonly str: string) {}

  compareTo(other: ComparableString): number {
    return this.str.localeCompare(other.str);
  }

  toString(): string {
    return this.str;
  }
}

File: haskell/Exercises.hs
module Exercises
  ( change,
    firstThenApply,
    powers,
    meaningfulLineCount,
    volume, 
    surfaceArea, 
    Shape(..),
  )
where

import Control.Exception (handle)
import Data.Char (isSpace)
import Data.List (find, isPrefixOf)
import Data.Map qualified as Map
import Data.Text (pack, replace, unpack)
import GHC.Data.ShortText (ShortText (contents))
import GHC.IO.FD (openFile)

change :: Integer -> Either String (Map.Map Integer Integer)
change amount
  | amount < 0 = Left "amount cannot be negative"
  | otherwise = Right $ changeHelper [25, 10, 5, 1] amount Map.empty
  where
    changeHelper [] remaining counts = counts
    changeHelper (d : ds) remaining counts =
      changeHelper ds newRemaining newCounts
      where
        (count, newRemaining) = remaining `divMod` d
        newCounts = Map.insert d count counts

firstThenApply :: [a] -> (a -> Bool) -> (a -> b) -> Maybe b
firstThenApply xs p f = f <$> find p xs

powers :: (Integral a) => a -> [a]
powers base = map (base ^) [0 ..]

meaningfulLineCount :: FilePath -> IO Int
meaningfulLineCount path = do
  contents <- readFile path
  let linesOfFile = lines contents
      validLines = filter isValidLine linesOfFile
  return (length validLines)

isValidLine :: String -> Bool
isValidLine line =
  not (null trimmedLine)
  && not (all isSpace trimmedLine)
  && not (startsWithHash trimmedLine)
  where
    trimmedLine = dropWhile isSpace line

startsWithHash :: String -> Bool
startsWithHash [] = False
startsWithHash (x : _) = x == '#'

data Shape
  = Sphere Double
  | Box Double Double Double
  deriving(Eq, Show)

surfaceArea :: Shape -> Double
surfaceArea (Sphere r) = 4.0 * pi * r * r
surfaceArea (Box l w h) = 2.0 * ((l*w) + (l*h) + (w*h))

volume :: Shape -> Double
volume (Sphere r) = (4.0 / 3.0) * pi * r * r * r
volume (Box l w h) = l * w * h

-- Write your binary search tree algebraic type here
